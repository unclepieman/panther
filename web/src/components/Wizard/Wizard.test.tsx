/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import { render, fireEvent } from 'test-utils';
import { useWizardContext, Wizard, WizardPanel } from './index';

describe('Wizard', () => {
  it('renders a step', () => {
    const { container } = render(
      <Wizard>
        <Wizard.Step title="Step Nickname">
          <WizardPanel>
            <WizardPanel.Heading title="Title" subtitle="Subtitle" />
            Content
            <WizardPanel.Actions>
              <WizardPanel.ActionNext>Continue</WizardPanel.ActionNext>
            </WizardPanel.Actions>
          </WizardPanel>
        </Wizard.Step>
      </Wizard>
    );

    expect(container).toMatchSnapshot();
  });

  it('renders a step nickname correctly', () => {
    const { getByText } = render(
      <Wizard>
        <Wizard.Step title="A">test</Wizard.Step>
        <Wizard.Step title="B">test</Wizard.Step>
      </Wizard>
    );

    expect(getByText('A')).toBeInTheDocument();
    expect(getByText('B')).toBeInTheDocument();
  });

  it("doesn't render a nickname if header is false", () => {
    const { queryByText } = render(
      <Wizard header={false}>
        <Wizard.Step title="A">test</Wizard.Step>
        <Wizard.Step title="A">test</Wizard.Step>
      </Wizard>
    );

    expect(queryByText('A')).not.toBeInTheDocument();
    expect(queryByText('B')).not.toBeInTheDocument();
  });

  it('renders the header correctly', () => {
    const { container, getByText } = render(
      <Wizard header={false}>
        <Wizard.Step>
          <WizardPanel>
            <WizardPanel.Heading title="Title" subtitle="Subtitle" />
          </WizardPanel>
        </Wizard.Step>
      </Wizard>
    );

    expect(container.querySelector('header')).toBeInTheDocument();
    expect(getByText('Title')).toBeInTheDocument();
    expect(getByText('Subtitle')).toBeInTheDocument();
  });

  it('renders the content of a step', () => {
    const { getByText } = render(
      <Wizard header={false}>
        <Wizard.Step>
          <WizardPanel>
            Content
            <WizardPanel.ActionNext />
          </WizardPanel>
        </Wizard.Step>
      </Wizard>
    );

    expect(getByText('Content')).toBeInTheDocument();
  });

  it('does not render back/next buttons by default', () => {
    const { queryByText, queryByAriaLabel } = render(
      <Wizard header={false}>
        <Wizard.Step>
          <WizardPanel>Content</WizardPanel>
        </Wizard.Step>
      </Wizard>
    );

    expect(queryByText('Next')).not.toBeInTheDocument();
    expect(queryByAriaLabel('Go Back')).not.toBeInTheDocument();
  });

  it('renders a back/next buttons when included', () => {
    const { queryByText, queryByAriaLabel } = render(
      <Wizard header={false}>
        <Wizard.Step>
          <WizardPanel>
            <WizardPanel.ActionPrev />
            <WizardPanel.ActionNext />
          </WizardPanel>
        </Wizard.Step>
      </Wizard>
    );

    expect(queryByText('Next')).toBeInTheDocument();
    expect(queryByAriaLabel('Go Back')).toBeInTheDocument();
  });

  it("allows overriding the next button's text", () => {
    const { queryByText } = render(
      <Wizard header={false}>
        <Wizard.Step>
          <WizardPanel>
            <WizardPanel.ActionNext>Continue</WizardPanel.ActionNext>
          </WizardPanel>
        </Wizard.Step>
      </Wizard>
    );

    expect(queryByText('Continue')).toBeInTheDocument();
  });

  it('correctly sets and resets the step Status', () => {
    const WizardStepStatusSetter = () => {
      const { setCurrentStepStatus } = useWizardContext();
      return <button onClick={() => setCurrentStepStatus('FAILING')}>Set Step Status</button>;
    };

    const WizardStepStatusDisplay = () => {
      const { currentStepStatus } = useWizardContext();
      return <div>{currentStepStatus.toLowerCase()}</div>;
    };

    const { queryByText, getByAriaLabel } = render(
      <Wizard header={false}>
        <Wizard.Step>
          <WizardPanel>
            A
            <WizardStepStatusDisplay />
            <WizardStepStatusSetter />
            <WizardPanel.ActionNext>Continue</WizardPanel.ActionNext>
          </WizardPanel>
        </Wizard.Step>
        <Wizard.Step>
          B
          <WizardStepStatusDisplay />
          <WizardStepStatusSetter />
          <WizardPanel.ActionPrev />
        </Wizard.Step>
      </Wizard>
    );

    expect(queryByText('pending')).toBeInTheDocument();

    fireEvent.click(queryByText('Set Step Status'));
    expect(queryByText('failing')).toBeInTheDocument();

    fireEvent.click(queryByText('Continue'));
    expect(queryByText('pending')).toBeInTheDocument();
    expect(queryByText('failing')).not.toBeInTheDocument();

    fireEvent.click(queryByText('Set Step Status'));
    expect(queryByText('failing')).toBeInTheDocument();

    fireEvent.click(getByAriaLabel('Go Back'));
    expect(queryByText('pending')).toBeInTheDocument();
    expect(queryByText('failing')).not.toBeInTheDocument();
  });

  it('prev/next and start over buttons work correctly', () => {
    const { getByText, queryByText, getByAriaLabel } = render(
      <Wizard header={false}>
        <Wizard.Step>
          <WizardPanel>
            A
            <WizardPanel.Actions>
              <WizardPanel.ActionNext />
            </WizardPanel.Actions>
          </WizardPanel>
        </Wizard.Step>
        <Wizard.Step>
          <WizardPanel>
            B
            <WizardPanel.Actions>
              <WizardPanel.ActionPrev />
              <WizardPanel.ActionNext />
            </WizardPanel.Actions>
          </WizardPanel>
        </Wizard.Step>
        <Wizard.Step>
          <WizardPanel>
            C
            <WizardPanel.Actions>
              <WizardPanel.ActionPrev />
              <WizardPanel.ActionNext />
            </WizardPanel.Actions>
          </WizardPanel>
        </Wizard.Step>
        <Wizard.Step>
          <WizardPanel>
            D
            <WizardPanel.Actions>
              <WizardPanel.ActionStart />
            </WizardPanel.Actions>
          </WizardPanel>
        </Wizard.Step>
      </Wizard>
    );

    expect(getByText('A')).toBeInTheDocument();

    fireEvent.click(getByText('Next'));
    expect(queryByText('A')).not.toBeInTheDocument();
    expect(getByText('B')).toBeInTheDocument();

    fireEvent.click(getByText('Next'));
    expect(queryByText('B')).not.toBeInTheDocument();
    expect(getByText('C')).toBeInTheDocument();

    fireEvent.click(getByAriaLabel('Go Back'));
    expect(getByText('B')).toBeInTheDocument();
    expect(queryByText('C')).not.toBeInTheDocument();

    fireEvent.click(getByAriaLabel('Go Back'));
    expect(getByText('A')).toBeInTheDocument();
    expect(queryByText('B')).not.toBeInTheDocument();

    fireEvent.click(getByText('Next'));
    fireEvent.click(getByText('Next'));
    expect(queryByText('C')).toBeInTheDocument();

    fireEvent.click(getByText('Next'));
    expect(queryByText('D')).toBeInTheDocument();

    fireEvent.click(getByText('Start Over'));
    expect(queryByText('D')).not.toBeInTheDocument();
    expect(queryByText('A')).toBeInTheDocument();
  });

  it('correctly go to a specific step', () => {
    const { getByText, queryByText } = render(
      <Wizard header={false}>
        <Wizard.Step>
          <WizardPanel>
            A
            <WizardPanel.Actions>
              <WizardPanel.ActionGoToStep stepIndex={2} />
            </WizardPanel.Actions>
          </WizardPanel>
        </Wizard.Step>
        <Wizard.Step>
          <WizardPanel>
            B
            <WizardPanel.Actions>
              <WizardPanel.ActionPrev />
              <WizardPanel.ActionNext />
            </WizardPanel.Actions>
          </WizardPanel>
        </Wizard.Step>
        <Wizard.Step>
          <WizardPanel>
            C
            <WizardPanel.Actions>
              <WizardPanel.ActionPrev />
              <WizardPanel.ActionNext />
            </WizardPanel.Actions>
          </WizardPanel>
        </Wizard.Step>
        <Wizard.Step>
          <WizardPanel>
            D
            <WizardPanel.Actions>
              <WizardPanel.ActionStart />
            </WizardPanel.Actions>
          </WizardPanel>
        </Wizard.Step>
      </Wizard>
    );

    expect(getByText('A')).toBeInTheDocument();

    fireEvent.click(getByText('Continue'));
    expect(queryByText('A')).not.toBeInTheDocument();
    expect(getByText('C')).toBeInTheDocument();
  });

  it('allows updating the context data correctly', () => {
    const WizardDataConsumer = () => {
      const { data } = useWizardContext();
      return (
        <div>
          {data.text}-{Object.keys(data).length}
        </div>
      );
    };

    const WizardDataSetter = () => {
      const { setData } = useWizardContext();
      return <button onClick={() => setData({ text: 'B' })}>Set Data</button>;
    };

    const WizardDataUpdater = () => {
      const { updateData } = useWizardContext();
      return <button onClick={() => updateData({ test: 'test' })}>Update Data</button>;
    };

    const WizardDataResetter = () => {
      const { resetData } = useWizardContext();
      return <button onClick={resetData}>Reset Data</button>;
    };

    const { getByText } = render(
      <Wizard initialData={{ text: 'A' }} header={false}>
        <Wizard.Step>
          <WizardPanel>
            <WizardDataConsumer />
            <WizardDataSetter />
            <WizardDataUpdater />
            <WizardDataResetter />
          </WizardPanel>
        </Wizard.Step>
      </Wizard>
    );

    expect(getByText('A-1')).toBeInTheDocument();

    fireEvent.click(getByText('Set Data'));
    expect(getByText('B-1')).toBeInTheDocument();

    fireEvent.click(getByText('Update Data'));
    expect(getByText('B-2')).toBeInTheDocument();

    fireEvent.click(getByText('Reset Data'));
    expect(getByText('A-1')).toBeInTheDocument();
  });

  it('correctly  resets everything when  `reset` is called', () => {
    const WizardDataConsumer = () => {
      const { data } = useWizardContext();
      return <div>{data}</div>;
    };

    const WizardDataSetter = () => {
      const { setData } = useWizardContext();
      return <button onClick={() => setData('B')}>Set Data</button>;
    };

    const WizardResetter = () => {
      const { reset } = useWizardContext();
      return <button onClick={reset}>Reset</button>;
    };

    const { getByText } = render(
      <Wizard initialData="A" header={false}>
        <Wizard.Step>
          <WizardPanel>
            First
            <WizardDataSetter />
            <WizardDataConsumer />
            <WizardPanel.Actions>
              <WizardPanel.ActionNext />
            </WizardPanel.Actions>
          </WizardPanel>
        </Wizard.Step>
        <Wizard.Step>
          <WizardPanel>
            Second
            <WizardResetter />
            <WizardDataConsumer />
          </WizardPanel>
        </Wizard.Step>
      </Wizard>
    );

    expect(getByText('A')).toBeInTheDocument();
    expect(getByText('First')).toBeInTheDocument();

    fireEvent.click(getByText('Set Data'));
    fireEvent.click(getByText('Next'));
    expect(getByText('B')).toBeInTheDocument();
    expect(getByText('Second')).toBeInTheDocument();

    fireEvent.click(getByText('Reset'));
    expect(getByText('A')).toBeInTheDocument();
    expect(getByText('First')).toBeInTheDocument();
  });
});
