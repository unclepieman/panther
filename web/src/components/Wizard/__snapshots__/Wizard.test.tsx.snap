// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wizard renders a step 1`] = `
@keyframes animation-0 {
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
}

.panther-12 {
  background-color: #23344e;
  border-radius: 4px;
  padding: 24px;
  margin-bottom: 24px;
  width: 100%;
  position: relative;
}

.panther-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding-top: 10px;
  margin-bottom: 60px;
  z-index: 2;
}

.panther-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  z-index: 2;
  opacity: 1;
}

.panther-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 25px;
  height: 25px;
  font-size: 0.75rem;
  font-weight: 700;
  border-radius: 99999px;
  border: 1px solid;
  border-color: #bdbdbd;
  background-color: transparent;
}

.panther-1 {
  font-size: 0.875rem;
  margin-left: 8px;
}

.panther-11 {
  padding-top: 12px;
}



.panther-9 {
  will-change: opacity;
  -webkit-animation: animation-0 300ms 0ms ease-in-out both;
  animation: animation-0 300ms 0ms ease-in-out both;
}

.panther-6 {
  margin-bottom: 40px;
  text-align: center;
}

.panther-4 {
  font-weight: 500;
  font-size: 1.25rem;
}

.panther-5 {
  font-size: 0.875rem;
  margin-top: 8px;
  color: #bdbdbd;
}

.panther-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  margin-top: 32px;
  margin-bottom: 16px;
}

.panther-7 {
  width: auto;
  outline: none;
  padding-left: 20px;
  padding-right: 20px;
  height: 46px;
  font-size: 0.938rem;
  -webkit-transition: border-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,background-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: border-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms,background-color 100ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  border-radius: 4px;
  border: 1px solid;
  border-color: #0081c5;
  background-color: #0081c5;
  cursor: pointer;
  color: #f6f6f6;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.panther-7:hover {
  background-color: #0a8bcf;
  border-color: #0a8bcf;
}

.panther-7:focus {
  background-color: #0a8bcf;
  border-color: #55d6ff;
}

.panther-7:active,
.panther-7[data-active=true] {
  background-color: #0077bb;
  border-color: #0077bb;
}

<div>
  <header
    id="main-header"
  />
  <article
    class="panther-12"
  >
    <ul
      class="panther-3"
    >
      <li
        class="panther-2"
      >
        <div
          class="panther-0"
        >
          1
        </div>
        <p
          class="panther-1"
        >
          Step Nickname
        </p>
      </li>
    </ul>
    <div
      class="panther-11"
    >
      <section
        class="panther-10"
      >
        <div
          class="panther-9"
        >
          <header
            class="panther-6"
          >
            <h1
              class="panther-4"
            >
              Title
            </h1>
            <p
              class="panther-5"
            >
              Subtitle
            </p>
          </header>
          Content
          <div
            class="panther-8"
          >
            <button
              class="panther-7"
              type="button"
            >
              Continue
            </button>
          </div>
        </div>
      </section>
    </div>
  </article>
  <footer
    id="footer"
  />
</div>
`;
